File: packages/apidash_mcp_core/src/graphql.ts
/**
 * APIDash CLI — GraphQL Executor
 * Sends GraphQL queries/mutations over HTTP POST.
 * Supports variables, operation names, and custom headers.
 */

import axios from "axios";
import { STATUS_REASONS } from "./data/api-data.js";

export interface GraphQLRequestContext {
  url: string;
  query: string;
  variables?: Record<string, unknown>;
  operationName?: string;
  headers?: Record<string, string>;
  timeoutMs?: number;
}

export interface GraphQLResponseData {
  url: string;
  status: number;
  statusText: string;
  headers: Record<string, string>;
  body: string;
  data?: unknown;
  errors?: unknown[];
  duration: number;
}

export interface GraphQLResult {
  success: boolean;
  data: GraphQLResponseData;
  errorMsg?: string;
}

export async function executeGraphQLRequest(ctx: GraphQLRequestContext): Promise<GraphQLResult> {
  const {
    url,
    query,
    variables,
    operationName,
    headers = {},
    timeoutMs = 30_000,
  } = ctx;

  const startTime = Date.now();

  const payload: Record<string, unknown> = { query };
  if (variables && Object.keys(variables).length > 0) payload.variables = variables;
  if (operationName) payload.operationName = operationName;

  const mergedHeaders = {
    "Content-Type": "application/json",
    Accept: "application/json",
    ...headers,
  };

  try {
    const response = await axios({
      method: "POST",
      url,
      headers: mergedHeaders,
      data: payload,
      timeout: timeoutMs,
      validateStatus: () => true,
    });

    const duration = Date.now() - startTime;
    const status = response.status;
    const statusText = STATUS_REASONS[status] ?? response.statusText ?? "";

    const raw =
      typeof response.data === "object"
        ? JSON.stringify(response.data, null, 2)
        : String(response.data ?? "");

    const parsedGql = typeof response.data === "object" ? response.data : null;

    return {
      success: true,
      data: {
        url,
        status,
        statusText,
        headers: response.headers as Record<string, string>,
        body: raw,
        data: parsedGql?.data,
        errors: parsedGql?.errors,
        duration,
      },
    };
  } catch (err: unknown) {
    const duration = Date.now() - startTime;
    const msg = err instanceof Error ? err.message : String(err);
    return {
      success: false,
      errorMsg: msg,
      data: {
        url,
        status: 0,
        statusText: "Network Error",
        headers: {},
        body: "",
        duration,
      },
    };
  }
}

File: packages/apidash_mcp_core/lib/src/graphql.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class GraphQLRequestContext {
  final String url;
  final String query;
  final Map<String, dynamic>? variables;
  final String? operationName;
  final Map<String, String>? headers;
  final int? timeoutMs;

  GraphQLRequestContext({
    required this.url,
    required this.query,
    this.variables,
    this.operationName,
    this.headers,
    this.timeoutMs,
  });
}

class GraphQLResponseData {
  final String url;
  final int status;
  final String statusText;
  final Map<String, String> headers;
  final String body;
  final dynamic data;
  final List<dynamic>? errors;
  final int duration;

  GraphQLResponseData({
    required this.url,
    required this.status,
    required this.statusText,
    required this.headers,
    required this.body,
    this.data,
    this.errors,
    required this.duration,
  });

  Map<String, dynamic> toJson() => {
    'url': url,
    'status': status,
    'statusText': statusText,
    'headers': headers,
    'body': body,
    'data': data,
    'errors': errors,
    'duration': duration,
  };
}

class GraphQLResult {
  final bool success;
  final GraphQLResponseData data;
  final String? errorMsg;

  GraphQLResult({
    required this.success,
    required this.data,
    this.errorMsg,
  });

  Map<String, dynamic> toJson() => {
    'success': success,
    'data': data.toJson(),
    if (errorMsg != null) 'errorMsg': errorMsg,
  };
}

Future<GraphQLResult> executeGraphQLRequest(GraphQLRequestContext ctx) async {
  final startTime = DateTime.now().millisecondsSinceEpoch;

  final payload = <String, dynamic>{
    'query': ctx.query,
  };
  if (ctx.variables != null && ctx.variables!.isNotEmpty) {
    payload['variables'] = ctx.variables;
  }
  if (ctx.operationName != null) {
    payload['operationName'] = ctx.operationName;
  }

  final mergedHeaders = <String, String>{
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    ...?ctx.headers,
  };

  try {
    final request = http.Request('POST', Uri.parse(ctx.url));
    request.headers.addAll(mergedHeaders);
    request.body = jsonEncode(payload);

    final timeout = Duration(milliseconds: ctx.timeoutMs ?? 30000);
    final streamedResponse = await request.send().timeout(timeout);
    final response = await http.Response.fromStream(streamedResponse);
    final duration = DateTime.now().millisecondsSinceEpoch - startTime;

    final status = response.statusCode;
    final statusText = response.reasonPhrase ?? "";
    
    dynamic parsedGql;
    String raw;
    try {
      parsedGql = jsonDecode(response.body);
      raw = JsonEncoder.withIndent('  ').convert(parsedGql);
    } catch (_) {
      raw = response.body;
      parsedGql = null;
    }

    dynamic responseData;
    List<dynamic>? errors;
    if (parsedGql is Map) {
      responseData = parsedGql['data'];
      if (parsedGql['errors'] is List) {
        errors = parsedGql['errors'];
      }
    }

    return GraphQLResult(
      success: true,
      data: GraphQLResponseData(
        url: ctx.url,
        status: status,
        statusText: statusText,
        headers: response.headers,
        body: raw,
        data: responseData,
        errors: errors,
        duration: duration,
      ),
    );
  } catch (err) {
    final duration = DateTime.now().millisecondsSinceEpoch - startTime;
    final msg = err.toString();
    return GraphQLResult(
      success: false,
      errorMsg: msg,
      data: GraphQLResponseData(
        url: ctx.url,
        status: 0,
        statusText: 'Network Error',
        headers: {},
        body: '',
        duration: duration,
      ),
    );
  }
}
